### PR TITLE
feat(wdio-prefix): introduce `wdio:{maxInstances,specs,exclude}` to WebdriverIO.Capabilities

### PR DESCRIPTION
**Pull request by vobu**
### feature
allow the capability config to contain a W3C-compliant `wdio` prefix for `maxInstances`,`specs`, and `exclude`:

```ts
const _config: Testrunner = {
    ...baseConfig,
    capabilities: [
        {
            browserName: 'safari',
            'wdio:maxInstances': 2,
            'wdio:specs': ['prefix-test-01.ts'],
            'wdio:exclude': ['prefix-test-02.ts']
        }
}
```

### type fix
When extending the `WebdriverIO.Capabilites` interface, the property `maxInstances` was missing. This resulted in an error when specifying `maxInstances` by using the extended type.

```typescript
// in some.types.ts
export interface wdi5Config extends WebdriverIO.Config {
      capabilities: wdi5Capabilities[] | wdi5MultiRemoteCapability
}
export interface wdi5Capabilities extends WebdriverIO.Capabilities {
    //...
}

// in wdio.conf.ts
export const config: wdi5Config = {
    // ...
    capabilities: [
            {
                maxInstances: 2, // <-- Error: Object literal may only specify known properties, and 'maxInstances' does not exist in type 'wdi5Capabilities'.ts(2353)
                browserName: "chrome",
                // ...
            }
    ]
}
```
## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] I have added proper type definitions for new commands (if appropriate)

### CODE REVIEW

File: packages/wdio-types/src/Capabilities.ts [hunk 1]
```diff
@@ -1130,10 +1134,10 @@ export interface BrowserStackCapabilities {
...
     debug?: boolean
     networkLogs?: boolean
-     /**
-     * https://www.browserstack.com/docs/app-automate/appium/debug-failed-tests/network-logs
-     * Enable viewing the response data in the Network Logs tab on your session
-     */
+    /**
+    * https://www.browserstack.com/docs/app-automate/appium/debug-failed-tests/network-logs
+    * Enable viewing the response data in the Network Logs tab on your session
+    */
```
Comment (christian-bromann) [inline packages/wdio-types/src/Capabilities.ts:1140]:
```suggestion
     * https://www.browserstack.com/docs/app-automate/appium/debug-failed-tests/network-logs
     * Enable viewing the response data in the Network Logs tab on your session
     */
```

File: packages/wdio-types/src/Capabilities.ts [hunk 2]
```diff
@@ -77,15 +77,19 @@ declare global {
...
             unhandledPromptBehavior?: string;
             /**
              * WebDriver clients opt in to a bidirectional connection by requesting a capability with the name "webSocketUrl" and value true.
              */
-            webSocketUrl?: boolean
+            webSocketUrl?: boolean;
+            /**
+             * Maximum number of total parallel running workers.
+             */
+            maxInstances?: number;
```
Comment (christian-bromann) [inline packages/wdio-types/src/Capabilities.ts:92]: Can we create a custom interface for WebdriverIO options in capabilities, have this interface extend from it and keep it as is since we only have true w3c capabilities here.

**Review by christian-bromann — COMMENTED** [review]
Thanks for the proposal, this makes me think of whether we should treat these "special" webdriverio capabilities more W3C conform. Maybe we should prefix them with `wdio:XXX`.

@vobu can you do that in the new interface? Afaik the following should be recognised: `maxInstances`, `specs`, `exclude`.

File: packages/wdio-types/src/Capabilities.ts (hunk 2 above)
Comment (vobu) [reply packages/wdio-types/src/Capabilities.ts:92]: right on, I'll give it a shot.

File: packages/wdio-types/src/Capabilities.ts [hunk 3]
```diff
@@ -39,9 +39,16 @@ export interface ProxyObject {
     noProxy?: string[];
 }
 
+export interface WdioOptions {
+    /**
+     * Maximum number of total parallel running workers (per capability)
+     */
+    maxInstances?: number;
+}
```
Comment (christian-bromann) [inline packages/wdio-types/src/Capabilities.ts:47]: Instead of redefining, better do this:
```suggestion
interface WDIOCapabilityOptions extends Pick<TestrunnerOptions, 'specs' | 'exclude' | 'maxInstances'> {}
```

File: packages/wdio-types/src/Capabilities.ts [hunk 4]
```diff
@@ -77,15 +84,15 @@ declare global {
             /**
              * Defines the current session’s strict file interactability.
              */
-            strictFileInteractability?: boolean,
+            strictFileInteractability?: boolean;
```
Comment (christian-bromann) [inline packages/wdio-types/src/Capabilities.ts:87]: Let's not use `;` they are not needed (not only here but also in other parts of the file.

**Comment by vobu** [comment]
you're too fast @christian-bromann :)
let me work on some more details, then I'll ping when done and ready for re-review

File: packages/wdio-types/src/Capabilities.ts (hunk 3 above)
Comment (vobu) [reply packages/wdio-types/src/Capabilities.ts]: done

File: packages/wdio-types/src/Capabilities.ts (hunk 4 above)
Comment (vobu) [reply packages/wdio-types/src/Capabilities.ts:117]: `,` → `;`?
this was done by the eslint-prettier setup and autoformat, not by me intentionally...

**Comment by vobu** [comment]
ok, now there should both be the `pick`ed type definitions:
- `wdio`-prefixed options 
- plain options marked `deprecated`

plus a test extending options per capability.

→ ping @christian-bromann :)

File: packages/wdio-types/src/Capabilities.ts (hunk 4 above)
Comment (christian-bromann) [reply packages/wdio-types/src/Capabilities.ts:117]: can we revert this?

File: packages/wdio-types/src/Capabilities.ts [hunk 5]
```diff
@@ -39,9 +40,45 @@ export interface ProxyObject {
     noProxy?: string[];
 }
 
+export interface CapabilityOptions extends Pick<TestrunnerOptions, 'specs' | 'exclude' | 'maxInstances'> {
```
Comment (christian-bromann) [inline packages/wdio-types/src/Capabilities.ts:43]: We don't need this if we define them in the interface with new docs.
```suggestion
export interface CapabilityOptions {
```

**Review by christian-bromann — COMMENTED** [review]
I am 💯 for using prefixed wdio caps, however we currently don't recognise them and need to make enhancements to the code.

File: packages/wdio-types/src/Capabilities.ts (hunk 4 above)
Comment (christian-bromann) [reply packages/wdio-types/src/Capabilities.ts:117]: yes, we should have a linter for this but that can be done in a different PR

Comment (vobu) [reply packages/wdio-types/src/Capabilities.ts]: done: all `;` removed

File: packages/wdio-types/src/Capabilities.ts (hunk 5 above)
Comment (vobu) [reply packages/wdio-types/src/Capabilities.ts]: done

**Comment by christian-bromann** [comment]
> however we currently don't recognise them and need to make enhancements to the code.

The code in question is:
- [596057](https://github.com/webdriverio/webdriverio/blob/d7e4b20dbaf354831eecd7a2298e79af94993207/packages/wdio-config/src/node/ConfigParser.ts#L171-L179)
- https://github.com/webdriverio/webdriverio/blob/c6db464b4b87b7e80b66eda4e3260b7be43b61a6/packages/wdio-cli/src/launcher.ts#L232

**Comment by christian-bromann** [comment]
> Or where do you see the `wdio:specs` and `wdio:exclude` being used?

No, this is exactly how I imagined. Let me know if you have further questions.

**Comment by vobu** [comment]
> > Or where do you see the `wdio:specs` and `wdio:exclude` being used?
> 
> No, this is exactly how I imagined. Let me know if you have further questions.

right on, then I'm on the right track 😄 
Further q's: 
- should the `wdio:specs` and `wdio:exclude` mix in with `config.specs` and `config.exclude` or overwrite?
- if so, which one takes precedence (if `wdio:exclude` excludes `$test-x`, but `config.specs` lists it)? 
  I'd naïvely assume "specific first", so `config.capabilty.wdio:{specs,exclude,maxInstances}` > `config.{specs,exclude,maxInstances}`

**Comment by christian-bromann** [comment]
> * should the `wdio:specs` and `wdio:exclude` mix in with `config.specs` and `config.exclude` or overwrite?

`wdio:specs` and `wdio:exclude` will take precedence and overwrite global configurations.

> * how should the case be handled when multiple `capabilities` list different files? Say:

In this case the first capability only runs `path.resolve('test-03.ts')` and the second only runs `path.resolve('test-05.ts')`

> I'd assume that this should result in "2 instances of Chrome run test-03.ts, but not test-04.ts. And neither test-05.ts nor test-06.ts".

No, this should initiate 1 Chrome and 1 Safari session. A `wdio:spec` defined in a capability does not effect other capabilities. Note that all this functionality is already there. You already define `spec` and `exclude` in the capabilities and it should work. What is needed is that we check for `spec` as well as `wdio:spec` capabilities.

**Comment by vobu** [comment]
hmmm, can't really do sth about the  ❌ e2e-mac ... might be worth a re-trigger (which I can't do due to repo perms)

**Comment by vobu** [comment]
and in terms of docs: wasn't sure whether to put any (yet) - if I should, maybe to here: https://webdriver.io/docs/capabilities#webdriverio-capabilities-to-manage-browser-driver-options ?

File: packages/wdio-config/src/node/ConfigParser.ts [hunk 6]
```diff
@@ -31,7 +31,9 @@ interface TestrunnerOptionsWithParameters extends Omit<Options.Testrunner, 'capa
 
 interface MergeConfig extends Omit<Partial<TestrunnerOptionsWithParameters>, 'specs' | 'exclude'> {
     specs?: Spec[]
+    'wdio:specs'?: Spec[]
     exclude?: string[]
+    'wdio:exclude'?: string[]
```
Comment (christian-bromann) [inline packages/wdio-config/src/node/ConfigParser.ts:36]: Can we revert changes in this file? These are options and not capabilities. We don't need to support prefixed options as we don't have to adhere to a standardized interface

File: packages/wdio-types/src/Capabilities.ts [hunk 7]
```diff
@@ -5,83 +5,119 @@ import type {
...
+    sslProxyPort?: number
+    socksProxy?: string
+    socksProxyPort?: number
+    socksVersion?: string
+    socksUsername?: string
+    socksPassword?: string
+    noProxy?: string[]
+}
+
+export interface CapabilityOptions {
```
Comment (christian-bromann) [inline packages/wdio-types/src/Capabilities.ts:42]: Can we remove this interface and merge everything into [`WebdriverIOCapabilities`](https://github.com/webdriverio/webdriverio/blob/0652d3c0212256108d3f505f98b95d97a5b5f591/packages/wdio-types/src/Capabilities.ts#L213)?

**Review by christian-bromann — COMMENTED** [review]
This looks great, I added some comments. Let's also document these caps in [`/website/docs/Capabilities.md`](https://github.com/webdriverio/webdriverio/blob/63ef7b125b33657f99558a7177c4d504f4cdb032/website/docs/Capabilities.md#L37)

**Comment by christian-bromann** [comment]
> if I should, maybe to here:

Yes, that looks like a good place!

File: packages/wdio-config/src/node/ConfigParser.ts (hunk 6 above)
Comment (vobu) [reply packages/wdio-config/src/node/ConfigParser.ts:36]: kk, done

File: packages/wdio-types/src/Capabilities.ts (hunk 7 above)
Comment (vobu) [reply packages/wdio-types/src/Capabilities.ts]: kk, done

**Comment by vobu** [comment]
added the [documentation](https://github.com/webdriverio/webdriverio/pull/11992/files#diff-f4f99dfb0ab78eb0d2ced19921eab3dd9d5337e4f98de51c2c6b83db41edf5baR64). 
yet CI fails due to a build error not specific for this commit (i hope 🤣 ).

File: packages/wdio-cli/src/launcher.ts [hunk 8]
```diff
@@ -271,7 +272,9 @@ class Launcher {
      * Format the specs into an array of objects with files and retries
      */
     private _formatSpecs(capabilities: (Capabilities.DesiredCapabilities | Capabilities.W3CCapabilities | Capabilities.RemoteCapabilities), specFileRetries: number) {
-        const files = this.configParser.getSpecs((capabilities as Capabilities.DesiredCapabilities).specs, (capabilities as Capabilities.DesiredCapabilities).exclude)
+        const specs = (capabilities as Capabilities.DesiredCapabilities).specs || (capabilities as CapabilityOptions)['wdio:specs']
+        const excludes = (capabilities as Capabilities.DesiredCapabilities).exclude || (capabilities as CapabilityOptions)['wdio:exclude']
```
[the remark this answers is not in the document]
Comment (christian-bromann) [reply packages/wdio-cli/src/launcher.ts:276]: I will create a ticket to simplify this in v9 but until then we have to deal with these options

File: packages/wdio-config/src/node/ConfigParser.ts [hunk 9]
```diff
@@ -169,12 +169,17 @@ export default class ConfigParser {
         this._config = deepmerge(this._config, object) as TestrunnerOptionsWithParameters
 
         /**
-         * overwrite config specs that got piped into the wdio command
+         * overwrite config specs that got piped into the wdio command,
+         * also adhering to the wdio-prefixes from a capability
          */
-        if (object.specs && object.specs.length > 0) {
+        if (object['wdio:specs'] && object['wdio:specs'].length > 0) {
```
Comment (christian-bromann) [inline packages/wdio-config/src/node/ConfigParser.ts:175]: This is why the build is failing, `object` are not capabilities here. We should revert these changes.

File: website/docs/Capabilities.md [hunk 10]
```diff
@@ -61,6 +61,24 @@ If the driver `binary` is set, WebdriverIO won't attempt to download a driver bu
 
 :::
 
+#### `wdio:maxInstances`
```
Comment (christian-bromann) [inline website/docs/Capabilities.md:64]: I think this is a bit misleading as these caps are not related to driver options, maybe the section could look like this:

```md
### WebdriverIO Capabilities to manage browser driver options

WebdriverIO defines a set of own capabilities to manage its automated session:

#### `wdio:chromedriverOptions`

Specific options passed into Chromedriver when starting it.

#### `wdio:geckodriverOptions`

Specific options passed into Geckodriver when starting it.

#### `wdio:edgedriverOptions`

Specific options passed into Edgedriver when starting it.

#### `wdio:safaridriverOptions`

Specific options passed into Safari when starting it.

#### `wdio:maxInstances`

Maximum number of total parallel running workers for the specific browser/capability. Takes precedence over [maxInstances](#configuration#maxInstances) and [maxInstancesPerCapability](configuration/#maxinstancespercapability).

Type: `number`

#### `wdio:specs`

Define specs for test execution for that browser/capability. Same as the [regular `specs` configuration option](configuration#specs), but specific to the browser/capability. Takes precedence over `specs`.

Type: `(String | String[])[]`

#### `wdio:exclude`

Exclude specs from test execution for that browser/capability. Same as the [regular `exclude` configuration option](configuration#exclude), but specific to the browser/capability. Takes precedence over `exclude`.

Type: `String[]`

WebdriverIO manages installing and running browser driver for you. WebdriverIO uses a custom capability that allows you to pass in parameters to the driver.

#### Common Driver Options
...
```

File: packages/wdio-cli/src/launcher.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (vobu) [reply packages/wdio-cli/src/launcher.ts:276]: again thanks for the explanation.
made the change you suggested, plus a small type cleanup in the `.map`

File: website/docs/Capabilities.md (hunk 10 above)
Comment (vobu) [reply website/docs/Capabilities.md:64]: kk, moved and expanded the section as you suggested.

File: packages/wdio-config/src/node/ConfigParser.ts (hunk 9 above)
Comment (vobu) [reply packages/wdio-config/src/node/ConfigParser.ts:175]: hmmm, but if the check for `wdio:{specs,exclude}` is removed, we don't get the merged config and the tests (e.g.) `https://github.com/webdriverio/webdriverio/pull/11992/files#diff-e81dd1eb85b53f80e2f3860c68a8712a3b1bc2fa846f47909facc4637baf17d1R616` fail.

Above is of course directly connected to the `MergeConfig` type def:
```ts
interface MergeConfig extends Omit<Partial<TestrunnerOptionsWithParameters>, 'specs' | 'exclude'> {
    specs?: Spec[]
    'wdio:specs'?: Spec[]
    exclude?: string[]
    'wdio:exclude'?: string[]
}
```
What to do? 😄

File: website/docs/Capabilities.md [hunk 11]
```diff
@@ -38,6 +38,43 @@ While the amount of fixed defined capabilities is very low, everyone can provide
...
+
+#### `wdio:exclude`
+
+Exclude specs from test execution for that browser/capability. Same as the [regular `exclude` configuration option](configuration#exclude), but specific to the browser/capability. Takes precedence over `exclude`.
+
+Type: `String[]`
+
+WebdriverIO manages installing and running browser driver for you. WebdriverIO uses a custom capability that allows you to pass in parameters to the driver.
+
+
```
Comment (christian-bromann) [inline website/docs/Capabilities.md:77]:
```suggestion
```

File: packages/wdio-config/src/node/ConfigParser.ts (hunk 9 above)
Comment (christian-bromann) [reply packages/wdio-config/src/node/ConfigParser.ts:177]: Oh ok, thanks for the call out. Let's leave it it as is then.

**Review by christian-bromann — APPROVED** [review]
One minor thing that I will manually merge. The rest LGTM 👍 

Thanks a lot!
